FlowCrypt/src/main/java/com/flowcrypt/email/jetpack/workmanager/HandlePasswordProtectedMsgWorker.kt

```diff
@@ -0,0 +1,425 @@
+ *         E-mail: DenBond7@gmail.com
+ */
+class HandlePasswordProtectedMsgWorker(context: Context, params: WorkerParameters) :
+  BaseMsgWorker(context, params) {
+  override suspend fun doWork(): Result = withContext(Dispatchers.IO) {
+    LogsUtil.d(TAG, "doWork")
+
+    if (isStopped) {
+      return@withContext Result.success()
+    }
+
+    try {
+      val account = AccountViewModel.getAccountEntityWithDecryptedInfoSuspend(
+        roomDatabase.accountDao().getActiveAccountSuspend()
+      ) ?: return@withContext Result.success()
+
+      val passwordProtectedCandidates = roomDatabase.msgDao().getOutboxMsgsByStatesSuspend(
+        account = account.email,
+        msgStates = listOf(MessageState.NEW_PASSWORD_PROTECTED.value)
+      )
+
+      if (passwordProtectedCandidates.isNotEmpty()) {
+        prepareAndUploadPasswordProtectedMsgsToFES(account)
+      }
+
+      return@withContext rescheduleIfActiveAccountWasChanged(account)
+    } catch (e: Exception) {
+      e.printStackTrace()
+      return@withContext Result.failure()
+    } finally {
+      LogsUtil.d(TAG, "work was finished")
+    }
+  }
+
+  private suspend fun prepareAndUploadPasswordProtectedMsgsToFES(account: AccountEntity) =
+    withContext(Dispatchers.IO) {
+      val apiRepository = FlowcryptApiRepository()
+      val keysStorage = KeysStorageImpl.getInstance(applicationContext)
+      val accountSecretKeys = PGPSecretKeyRingCollection(keysStorage.getPGPSecretKeyRings())
+
+      var list: List<MessageEntity>
+      var lastMsgUID = 0L
+      while (true) {
+        list = roomDatabase.msgDao().getOutboxMsgsByStatesSuspend(
+          account = account.email,
+          msgStates = listOf(MessageState.NEW_PASSWORD_PROTECTED.value)
+        )
+
+        if (list.isEmpty()) {
+          break
+        }
+
+        val iterator = list.iterator()
+        var msgEntity: MessageEntity? = null
+
+        while (iterator.hasNext()) {
+          val tempMsgDetails = iterator.next()
+          if (tempMsgDetails.uid > lastMsgUID) {
+            msgEntity = tempMsgDetails
+            break
+          }
+        }
+
+        if (msgEntity == null) {
+          msgEntity = list[0]
+        }
+
+        lastMsgUID = msgEntity.uid
+
+        try {
+          if (msgEntity.isEncrypted == true && msgEntity.isPasswordProtected) {
+            //get msg attachments(including forwarded that can be encrypted)
+            val attachments = roomDatabase.attachmentDao().getAttachmentsSuspend(
+              account = msgEntity.email,
+              label = JavaEmailConstants.FOLDER_OUTBOX,
+              uid = msgEntity.uid
+            ).map {
+              it.copy(
+                forwardedFolder = "Outbox",
+                forwardedUid = Random.nextLong(),
+                decryptWhenForward = true
+              )
+            }
+
+            //create MimeMessage from content + attachments
+            val mimeMsgWithAttachments = EmailUtil.createMimeMsg(
```

Comment:
Let's call this `plainMimeMsgWithAttachments`? Is that what it is?

Reply:
> Is that what it is?

You are right